Image resize cloud function

radhikamattoo opened the pull request:
This is the implementation of the image resizing cloud function based on this [youtube video](https://www.youtube.com/watch?v=ZIe0Fn9OtnY&feature=youtu.be)

I have added the testing directory & file, but wanted to get feedback first in case I accidentally delete important files while testing the function. 

Radhika Mattoo

Commit:
created skeleton resize function

Commit:
added test directory

Commit:
changed to cloud storage instead of db, added node modules

Commit:
implemented image resizing using image magick

Commit:
simplified code based on youtube example

radhikamattoo commented:
I can't add you as a reviewer myself @apsdehal

yoav-zibin commented on functions/test/testing.js:
```diff
@@ -0,0 +1 @@
+var myFunctions = require('../index');
```
What's the purpose of this file? Right now it has no tests :)
Remove it for now (you can add it if you add tests)

yoav-zibin commented on package.json:
```diff
@@ -24,42 +23,16 @@
+      "version": "https://registry.npmjs.org/@google-cloud/common/-/common-0.13.6.tgz"
     },
     "@google-cloud/functions-emulator": {
-      "version": "https://registry.npmjs.org/@google-cloud/functions-emulator/-/functions-emulator-1.0.0-alpha.25.tgz",
```
What's the reason for all the changes here?
If you revert this file (undo all changes), will your cloud function work?

yoav-zibin commented on functions/index.ts:
```diff
@@ -245,3 +246,26 @@ functions.database.ref('gamePortal/gameSpec/reviews/{reviewedGameSpecId}/{review
+
+exports.resizeImage =
+functions.storage.object().onChange.onChange(async (event:any) => {
+  console.log("Resizing the image");
```
Move the console.log 3 lines down, and also log the path, bucket, imgType, uid, id, etc :)

yoav-zibin commented on functions/index.ts:
```diff
@@ -245,3 +246,26 @@ functions.database.ref('gamePortal/gameSpec/reviews/{reviewedGameSpecId}/{review
+  const bucket = event.data.bucket;
+  const [imgType, uid, id] = path.split("/");
+
+  if(imgType !== "upload") return;
```
do:
return null;
(or else you'll see warnings in firebase logs)

yoav-zibin commented on functions/index.ts:
```diff
@@ -245,3 +246,26 @@ functions.database.ref('gamePortal/gameSpec/reviews/{reviewedGameSpecId}/{review
+
+  // Resize it and upload it to cloud storage
+  console.log("Resizing image and uploading to cloud storage...");
+  await spawn('convert', [tmpFilePath, '-thumbnail', '900x900>', tmpFilePath]);
```
Add a comment about what this convert does:
it calls the function convert from imagemagick and pass it some parameters, like -thumbnail , see doc here:
http://www.imagemagick.org/Usage/thumbnails/

apsdehal commented:
Only professor has access to this one. @yoav-zibin can give both you and me access to the repo.

yoav-zibin commented on functions/index.ts:
```diff
@@ -245,3 +246,26 @@ functions.database.ref('gamePortal/gameSpec/reviews/{reviewedGameSpecId}/{review
+  await file.download({ destination : tmpFilePath});
+
+  // Resize it and upload it to cloud storage
+  console.log("Resizing image and uploading to cloud storage...");
```
Let's generate 3 images:
1) An image of exactly the same size and similar type (type is either png/jpg), but using quality=70 (see about quality: http://www.imagemagick.org/script/command-line-options.php?#quality )
2) Same as 1, with quality=50
2) A thumbnail image of lower size (at most 50x50), and always with type=JPG

We can't always use jpg, because for JPG: "It also does NOT allow any form of transparency." (see http://www.imagemagick.org/Usage/thumbnails/)
And if the input is a png, we need to preserve transparency. 

Add comments to the function saying what's the input and what's the output.
E.g., given that a user uploaded an image to:
gs://universalgamemaker.appspot.com/images/-KtgKdhyzZ6HrXsGwcpi.png
we create images in:
gs://universalgamemaker.appspot.com/quality70/-KtgKdhyzZ6HrXsGwcpi.png
gs://universalgamemaker.appspot.com/quality50/-KtgKdhyzZ6HrXsGwcpi.png
gs://universalgamemaker.appspot.com/thumbnail/-KtgKdhyzZ6HrXsGwcpi.jpg

yoav-zibin commented on functions/index.ts:
```diff
@@ -1,12 +1,13 @@
 'use strict';
```
Nice! Happy to see your first pull request, good job :)

Did you try it on your own firebase project?
Try it with a couple of images (both JPG and PNG) of different sizes (as big as 1024x1024, and small as 50x50), and visually look at the results (in different quality options; see my comments below).
Email me some screenshots :)

Commit:
generating 3 new files for each file uploaded and reverted package.json

Commit:
switched order of resizings, debugging imagemagick

Commit:
deleted unused test dir

Commit:
removed debug log

radhikamattoo commented:
Hi @yoav-zibin I've updated my code:
- Reverted the `package.json` to the last commit from your `gh-pages` branch
- Deleted the unused `test` dir 
- The youtube video's code wasn't working for me (kept throwing errors) so I decided to follow an example file on firebase's github [here](https://github.com/firebase/functions-samples/blob/master/convert-images/functions/index.js)

I'm still having trouble with the image resizing using ImageMagick. The docs are kind of confusing, so I though that the array of arguments being passed to spawn would be similar to the `-thumbnail` method, like this: 

`return spawn('convert', [tempFilePath, '-quality', '50', temp50]);`

However, when I test it by uploading an image to my test project on firebase, all the new images are the same thumbnail instead. I convert the images to different paths each time I call `spawn`, but for some reason it's not working.  Other than that, I need to convert the thumbnail to a JPG, but I wanted to fix this error first. Thanks for any help!

yoav-zibin commented:
Try debugging it a bit more :)
So you always get the same thumbnail, right? So probably tempFilePath is
always the same value.
What do you see in the logs?


+            console.log("Download path: ", *tempFilePath*);
+            console.log("File50 path: ", filePath50);
+            console.log("File70 path: ", filePath70);
+            console.log("Thumbnail path: ", thumbnailPath);
+            console.log("Resizing the image with path: ", filePath, " and
filename: ", fileName);

On Mon, Feb 5, 2018 at 8:10 PM, Radhika Mattoo <notifications@github.com>
wrote:

> Hi @yoav-zibin <https://github.com/yoav-zibin> I've updated my code:
>
>    - Reverted the package.json to the last commit from your gh-pages
>    branch
>    - Deleted the unused test dir
>    - The youtube video's code wasn't working for me (kept throwing
>    errors) so I decided to follow an example file on firebase's github
>    here
>    <https://github.com/firebase/functions-samples/blob/master/convert-images/functions/index.js>
>
> I'm still having trouble with the image resizing using ImageMagick. The
> docs are kind of confusing, so I though that the array of arguments being
> passed to spawn would be similar to the -thumbnail method, like this:
>
> return spawn('convert', [tempFilePath, '-quality', '50', temp50]);
>
> However, when I test it by uploading an image to my test project on
> firebase, all the new images are the same thumbnail instead. I convert the
> images to different paths each time I call spawn, but for some reason
> it's not working. Other than that, I need to convert the thumbnail to a
> JPG, but I wanted to fix this error first. Thanks for any help!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/yoav-zibin/FirebaseExample/pull/1#issuecomment-363276212>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzZHcOHRDqgMGdR_PAp-rPY6pL1mQotks5tR6aWgaJpZM4R6Hbi>
> .
>

Commit:
reordered if statements for proper exit of function

Commit:
fixed filename variables to force thumbnails to be jpgs

apsdehal commented:
Is this PR abandoned? Can we close this?
